cognite/client/data_classes/data_modeling/instances.py

```diff
@@ -1446,3 +1448,305 @@ def __delitem__(self, key: str | tuple[str, str] | tuple[str, str, str]) -> None
             del self[key[0]][key[1]][key[2]]
         else:
             raise ValueError(f"Invalid key: {key}")
+
+
+class PropertyOptions:
+    """This is a descriptor class for instance properties in a typed class.
+
+    It is used when you have a property that has a different name in the Data Model
+    compared to the name in the Python class.
+
+    Args:
+        identifier (str | None): The name of the property in the Data Model. Defaults to the name of the property in the Python class.
+    """
+
+    def __init__(self, identifier: str | None = None) -> None:
+        self.name = cast(str, identifier)  # mypy help, set_name guarantees str
+
+    def __set_name__(self, owner: type, name: str) -> None:
+        self.name = self.name or name
+        if self.name in _RESERVED_PROPERTY_NAMES:
+            self.name = f"{_RESERVED_PROPERTY_CONFLICT_PREFIX}{self.name}"
+
+    def __get__(self, instance: Any, owner: type) -> Any:
+        try:
+            return instance.__dict__[self.name]
+        except KeyError:
+            raise AttributeError(f"'{owner.__name__}' object has no attribute '{self.name}'")
+
+    def __set__(self, instance: Any, value: Any) -> None:
+        try:
+            instance.__dict__[self.name] = value
+        except KeyError:
+            raise AttributeError(f"'{instance.__class__.__name__}' object has no attribute '{self.name}'")
+
+    def __delete__(self, instance: Any) -> None:
+        try:
+            del instance.__dict__[self.name]
+        except KeyError:
+            raise AttributeError(f"'{instance.__class__.__name__}' object has no attribute '{self.name}'")
+
+
+class TypedInstance(ABC):
+    @classmethod
+    @abstractmethod
+    def get_source(cls) -> ViewId:
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def _base_properties() -> set[str]:
+        raise NotImplementedError
+
+    def _dump_properties(self) -> dict[str, Any]:
+        props = {key: prop for key, prop in vars(self).items() if key not in _RESERVED_PROPERTY_NAMES}
+        return _PropertyValueSerializer.serialize_values(props, camel_case=True)
+
+    @classmethod
+    def _load_instance(cls, resource: dict[str, Any], properties: dict[str, Any]) -> Self:
+        kwargs: dict[str, Any] = {}
+        signature = inspect.signature(cls.__init__)
+        kwargs.update(cls._load_properties(properties, signature))
+        kwargs.update(cls._load_base_properties(resource, signature))
+        return cls(**kwargs)
+
+    @classmethod
+    def _load_base_properties(cls, resource: dict[str, Any], signature: inspect.Signature) -> dict[str, Any]:
+        args: dict[str, Any] = {}
+        for key in cls._base_properties():
+            camel_key = to_camel_case(key)
+            if camel_key in resource:
+                if key in signature.parameters and key in signature.parameters:
```

Comment:
```suggestion
                if key in signature.parameters:
```

Reply:
Twice as safe to cosmic rays got to be worth something 😂